Frame Annotation Tooling

abdullah-ukaea opened the pull request:
This PR introduces frame-based annotation for UFO image data and integrates it with existing UI tooling and the backend annotation APIs.

We currently support rectangle-only annotations (polygons to follow in a later PR) with per-instance tracking using track_id.

Addresses issue https://github.com/ukaea/toktagger/issues/32 

# What this adds

**Frame-level annotation view for UFO tasks**
* Adds a new UFOView that wraps FrameView for per-frame annotation.
* Renders frames using base64-encoded image data from the backend.
* Frame navigation via Prev / Next / Jump wired through the existing dataParams.frame plumbing.

**Rectangle-only Annotorious integration with tracking**
* Uses @annotorious/react to draw rectangles only.
* Each annotation is stamped with class label + track_id to enable instance-style tracking.
* Drawing is enabled only when a selected instance profile (class + track) exists in the UFO toolbar.

**Per-frame localStorage persistence + aggregation**
* Per-frame annotations are stored as W3C image annotations in localStorage (keyed by app://p/<project>/s/<sample>/f/<frame>).
* A background auto-save loop flushes edits from Annotorious into localStorage.
* window.ufoCollectForSave() sweeps all frames for a sample and returns a flat W3C list for saving.

Toolbar integration (class + instance profiles)
* Left-hand UFO toolbar provides:
     - Class picker (hard-coded label map for now)
     - Instance panel listing (class_name, track_id) profiles
     - Cross-frame count badges per instance
* Toolbar + FrameView stay in sync via window.* globals + ufo:state events.
* Cross-frame instance counts are computed with a chunked localStorage scanner (scanInstanceCountsChunked).

**Backend save + navigation wiring**
* UFO Save converts: W3C → COCO frames → VideoBoundingBox[] and PUTs to: PUT /projects/{project_id}/samples/{sample_id}/annotations (API on :8002)
* Important: backend expects created_by in AnnotatorTypes; for UI annotations we send created_by: "manual".
* Next sample: saves current UFO payload then loads next sample via: GET /projects/{project_id}/samples/next
* Jump to shot: saves multi-frame annotations before navigating to the target sample from shot_id.

**Bulk delete & clear flows**
* Right-click an instance → delete that instance’s annotations across all frames (with counts preview).
* Delete All Instances → wipes all instance profiles and annotations for the sample across every frame.
* Toolbar buttons delegate to frame-level helpers:
   - ufoClearAllFrames() — removes all frame keys for the sample from localStorage
   - ufoClearCurrent() — clears current frame overlay + storage


# Implementation notes

* Annotorious ImageAnnotator is used as the drawing surface; tool is restricted to rectangles.
* An imperative AnnoBridge wraps useAnnotator() and:
   - Normalizes rectangle annotations via normalizeWithMode, stamping class/track metadata from toolbar selection.
   - Tracks dirty state and exposes: persistWorkingNow, hydrateOverlay, clearOverlaySilently, hasUnsaved, markSaved.
   - Listens to create/update/delete events and re-normalizes overlay.
   - Supports “auto quick add”: if a second shape for the same instance is drawn in a frame, a new profile can be created and the duplicate is retagged.

# Files added / modified

* toktagger/ui/src/app/frames/components/adapters.ts
* toktagger/ui/src/app/frames/components/bridge.tsx
* toktagger/ui/src/app/frames/components/frames.tsx
* toktagger/ui/src/app/frames/components/lib.ts
* toktagger/ui/src/app/frames/components/ui.tsx
* toktagger/ui/src/app/components/tools/toolbar.tsx
* toktagger/ui/src/app/projects/project_id/samples/sample_id/page.tsx

# Future Work

* Add polygon tool support (alongside rectangles)
* Pan/zoom/drag UX for the image view
* Timeline/transport controls for faster navigation through frames
* Deeper backend integration (e.g., loading existing stored frame annotations into localStorage/hydration path
* Improve jump-to-frame UX (direct frame number input + enter)

abdullah-ukaea committed:
ui: add annotorious + openseadragon dependencies for frame annotation

abdullah-ukaea committed:
scripts/setup: add create_image_samples and seed UFO image samples from /data/jet_images

abdullah-ukaea committed:
added 1 line to docker compose to mount jet images

abdullah-ukaea committed:
renamed files

abdullah-ukaea committed:
updated import

abdullah-ukaea committed:
added in adapter, bridge, lib and ui helper files in frames/components

abdullah-ukaea committed:
updated frames.tsx with frame integration tooling draft 1

abdullah-ukaea committed:
updated toolbar script, wrong filenames and imports

abdullah-ukaea committed:
renamed adapters file

abdullah-ukaea committed:
fixed toolbar ui

abdullah-ukaea committed:
updated annotation functionality

abdullah-ukaea committed:
fixed toolbar class labels, instance profiles

abdullah-ukaea committed:
latest stable version of frame annotation tooling integration

abdullah-ukaea committed:
Add proper UFO bbox schemas to toktagger/ui/src/types.ts and Make the backend payload type in lib.ts require track_id: string

abdullah-ukaea committed:
Patch VideoBoundingBox type + generator to include type and created_by

abdullah-ukaea committed:
fixed ui misalignment in toolbar

abdullah-ukaea committed:
fixed type error

abdullah-ukaea committed:
added a few lines to gitignore

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Refactor toolbar.tsx to satisfy ESLint without changing behavior. Fix conditional hook usage by removing early returns and returning null after hooks when non-UFO data is missing. Improve type safety by adding typed UFO window/event contracts and removing any from UFO state handling and conversion paths.

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Refactor  to remove all explicit  types by replacing them with  plus small type guards and a minimal typed Annotorious API surface. This keeps the rectangle-only behavior and normalization logic the same while making the bridge’s imperative handle and event subscriptions type-safe and ESLint-clean.

abdullah-ukaea committed:
Updated frames.tsx to remove all explicit any usage by adding typed window helpers, small type guards, and typed CustomEvent details without changing runtime behavior. We also fixed the Next.js <img> lint warning with a targeted eslint disable and removed the AnnoBridge ref cast so the file passes lint cleanly.

abdullah-ukaea committed:
Refactor lib.ts to remove any by using unknown plus small type guards and helpers for safer parsing of annotation and localStorage data. This keeps behavior the same while improving type-safety across label extraction, cross-frame scanning, and export conversion utilities.

abdullah-ukaea committed:
Refactor SamplePage to remove any usage by switching to unknown with small type guards for safer property access. This improves type safety in frame navigation and error handling while keeping the page behavior unchanged and satisfying ESLint.

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Ran the command npx prettier --write . to run pretter for formatting

abdullah-ukaea committed:
added a line to prettier ignore to ignore a vite cached folder

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Delay UFO instance profile creation until the first box is drawn instead of on class dropdown selection. Update drawing enablement so selecting a class arms drawing even when no instance exists yet. On first annotation creation, automatically create/select a new instance profile and stamp the annotation to it, and adjust the empty-state UI text to reflect the new workflow.

abdullah-ukaea committed:
Remove the confusing Clear ALL button from the UFO toolbar so users don’t accidentally clear boxes while keeping instance profiles. Rename Clear Current to Clear Current Frame and center it to make the primary per-frame action more obvious. Keep Delete All Instances as the only global wipe option.

abdullah-ukaea committed:
Prevent UFO instance selection from sticking when you move between frames. Added a ufoSelectionSource flag so only instances the user explicitly clicks stay selected across navigation. Auto-created selections are marked as auto and are cleared on Prev/Next/Jump while keeping the class armed, so new boxes default to creating a new instance.

abdullah-ukaea committed:
Start with frame number 200 instead of 292 when loading an annotation session

abdullah-ukaea committed:
Delete the unused function, since nothing calls this anymore. Deprecated function removal in toolbar.tsx

abdullah-ukaea committed:
Delete unused import

abdullah-ukaea committed:
Prettier formmating to pass the linter

Commit:
chore: update build output [skip ci]

wk9874 committed:
Change image data params to accept None and catch in data loader

abdullah-ukaea committed:
updated samples page to start at frame null as discussed instead of start frame 200 or 292 now that the backend has been updated in previous commit

Commit:
chore: update build output [skip ci]

wk9874 committed:
Sort frame paths and take first one

jblake42 commented on docker-compose.dev.yml:
```diff
@@ -42,6 +42,7 @@ services:
       - mongo
     volumes:
       - ./data/:/app/data
+      - ./data/:/data
```
Is this definitely needed? It seems weird we are mapping this host volume twice for this service. The `cwd` of the service is `/app` and looking at the dataloader this is where it looks for locally stored files/directories. I could be wrong though, I've only glanced over the backend code

jblake42 replied:
I think you may have already corrected this

abdullah-ukaea replied:
I agree, addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/0c9888c27deb3bf3dd1895646a1e742b2101a2e5

jblake42 commented on toktagger/api/core/data_loaders.py:
```diff
@@ -105,8 +105,10 @@ def get_sample(self, shot_id: int, sample_data: FileData) -> ImageData:
+        if self.params.name != "image":
+            raise ValueError("Must provide image data parameters!")
+        elif self.params.frame is None:
+            file_path = sorted(dir_path.iterdir())[0]
```
Might be an edge case but this looks unsafe to me - do we risk an `IndexError` exception here if an empty directory is provided?

abdullah-ukaea replied:
@wk9874

wk9874 replied:
Good shout

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
This toolbar file is getting pretty nasty with lots of conditional logic etc. I think we need to look at abstracting this so that projects can load in their own components when needed rather than making this massive file that does different things depending on the project task.

Happy for this to be made an issue and we look at it outside the PR

abdullah-ukaea replied:
Will try to address in the refactor PR

jblake42 replied:
Created an issue

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
-      console.warn("ELM data is not available");
-      return;
-    }
+  if (!isUfo) {
```
This is the sort of conditional logic I am talking about - it makes the code very hard to follow. This isn't your fault, but I think we need to give this a bit of thought as I don't think this is the way personally

abdullah-ukaea replied:
Will try to address in the refactor PR - future work

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
There also seem to be comments in here that look like debugging/tracking changes - these should not be left in

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/2f04a0c2c3772f6ad4a004823e33894581ed9b18

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+    const onState = (e: Event) => {
+      const detail = (e as CustomEvent<unknown>).detail;
+      const d = (
+        detail && typeof detail === "object" ? (detail as UfoStateDetail) : {}
```
The internal type assertion is not needed as you do it again to the whole ternary result.

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/2218d66ef788c9e5bd7674f7adf6593b11f72ccd

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+  );
+
+  /**
+   * Listen to FrameView state events so toolbar stays in sync (instances, selection, counts, registry).
```
I am not the biggest fan of using these custom events to keep things synchronised. I feel like this would be a good use for contexts like I used for the other tooling. Since the toolbar is a fairly high level it may be a challenge as to where we put the context, but I can't see it causing a massive issue. This does work, I just worry about if this adds complexity and strange global variables as opposed to the tested context approach.

jblake42 replied:
The major issue I see with relying on an event system is that if an event is missed / a component loads after a component is loaded then it breaks the synchronisation

abdullah-ukaea replied:
Will try to address in the refactor PR / future work

abdullah-ukaea committed:
This change makes UFO frame annotations load correctly from the backend by converting the saved video_bounding_box records into W3C rectangle annotations and hydrating Annotorious so boxes draw immediately on page load. We also rebuild the left-side Instances list by scanning the stored W3C annotations, so it doesn’t come back empty after a refresh. Finally, we added a per-sample working session dirty marker that is set on local autosave and cleared after a successful backend save, so we can prefer backend data unless there’s unsaved local work.

abdullah-ukaea committed:
eslinter patch to replace any with DOM type PerformanceNavigationTiming

abdullah-ukaea committed:
eslinter patch to replace any usage to unkown type and replace as any castg with as unkown as ImageAnnotation

abdullah-ukaea committed:
prettier linter patch

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
quick formmating

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+        }
+      }
+
+      if ("selectedClassName" in d) {
```
I think the `nullish coalescing operator` means we do not really need the outer conditional check

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/b92b2018bf14509efb22c2567594bbc6f2c10540

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+
+    const keyPrefix = "anno::w3c::" + `app://p/${project_id}/s/${sample_id}/`;
+
+    let stopScan: (() => void) | null = null;
```
I'm a little bit confused about the whole start/stop thing - could you explain what it does please?

abdullah-ukaea replied:
The “start/stop” pattern is just there to manage a chunked async scan safely. scanInstanceCountsChunked walks localStorage in small setTimeout(0) chunks and returns a stopScan() cancel function; we call stopScan() before starting a new scan (and on unmount) so we don’t end up with multiple overlapping scans racing to call setInstanceCounts with stale/partial results while a new scan is already in progress. The setInterval simply re-triggers a fresh scan every second so the instance profile count badges stay accurate as FrameView writes/clears annotations in localStorage.

abdullah-ukaea replied:
added some comments to help with readability of this section https://github.com/ukaea/toktagger/pull/137/commits/0c22d56fcc59dba06d0cc0bb854288aaaff258e9 I plan to add more comments to all the files when i am at the end of this code review

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -44,6 +46,55 @@ import { JumpDetectionTool } from "../annotators/jump";
+  profileCounts?: Record<string, number>;
+};
+
+declare global {
```
Overwriting the Window type is a little confusing especially in the global scope - if there is a way to avoid it I would. There may not be though

abdullah-ukaea replied:
Will try to address in the refactor PR / future work

abdullah-ukaea committed:
reverted a data mount to the docker compose file back to what main branch uses

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+    const cocoFrames = w3cToCocoFrames(w3cList, true);
+    const videoBBoxes = cocoFramesToVideoBBoxes(cocoFrames);
+
+    return videoBBoxes as unknown as Annotation[];
```
How come you cast via `unknown` is it a linter thing?

abdullah-ukaea replied:
not sure to be honest, its not a linter thing Direct as Annotation[] compiles cleanly, so I’ve removed the unknown cast in https://github.com/ukaea/toktagger/pull/137/commits/80cbd99544f2e96ca7d092a08b1afde5faf05060

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+  const handleUfoSave = async () => {
+    try {
+      const payload = await collectUfoPayloadForBackend();
+      const response = await saveAnnotations(project_id, sample_id, payload);
```
`project_id` can be `undefined`

abdullah-ukaea replied:
addressed in commit http://github.com/ukaea/toktagger/pull/137/commits/9b1b0f92952e9f1449ebedb5ad2c0e3553fe44ff

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+  const handleUfoNextSample = async () => {
+    try {
+      const payload = await collectUfoPayloadForBackend();
+      await saveAnnotations(project_id, sample_id, payload);
```
Same for this `project_id` and the one below

jblake42 replied:
I won't comment everyone but there are quite few more below

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/880b2532ab2966efdd9b37276311d3c13fc60927

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+    };
+
+    return (
+      <SearchField
```
On my system I cannot input any data into the box - it looks as if it is immediately overwritten with a blank string

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/eceb97165e38862d71c96e06111fac1f2c763039

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
-          gap="size-100"
-          width="100%"
-        >
+      {isUfo ? (
```
Again, this is the place where if we put some work into how we best compose this component it would make it so much clearer imo - we wouldn't need layers of conditions

abdullah-ukaea replied:
future work / refactor PR

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+            <div className="max-w-[16rem] mx-auto mb-4">
+              <div className="mb-2">
+                <Flex gap="size-100" alignItems="center" wrap>
+                  <Button
```
This button component is malformed - missing `variant`. Also, do we need this? It looks a bit out of place?

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/1614f2b9083b7a77c672cf5e3f912ce25dbe77dd

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+    return null;
+  }
 
   return (
```
The UFO toolbar is not very light mode friendly

abdullah-ukaea replied:
Will try to address in the refactor PR / future work

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+              <div className="mb-2">
+                <Flex gap="size-100" alignItems="center" wrap>
+                  <Button
+                    isQuiet
```
This is depracated

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/1614f2b9083b7a77c672cf5e3f912ce25dbe77dd

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+                  wrap
+                >
+                  <Button
+                    isQuiet
```
Deprecated and teh component is missing `variant`

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/1614f2b9083b7a77c672cf5e3f912ce25dbe77dd

jblake42 commented on toktagger/ui/src/app/components/tools/toolbar.tsx:
```diff
@@ -419,195 +466,655 @@ export default function ToolBar({
+                    onPress={async () => {
+                      try {
+                        await ufoClearCurrent();
+                        ToastQueue.positive("Cleared current frame.", {
```
I'm unsure if we need positive feedback for events that provide immediate feedback to the user - for example it is clear that the action happened because the annotations dissappear. Realistically the toast is discreet enough that it wouldn't be an annoyance

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/e7f02c8e7718e2f95c3610dc41c72f18fc2942bf

jblake42 commented on toktagger/ui/src/app/frames/components/adapters.ts:
```diff
@@ -0,0 +1,63 @@
+ * Storage layout:
+ *   localStorage["anno::w3c::" + sourceKey] = JSON.stringify(ImageAnnotation[])
+ */
+export function W3CImageFormat(sourceKey: string) {
```
Nice adapter pattern

jblake42 reviewed (commented):
I've made it though some of the files but still have a way to go. Pushing these now so they don't get lost. I think the local storage approach you have taken is very clever, I just wonder if a more standardised approach to align to other tooling could be simpler? I haven't really gotten into depth yet so I am far from confident in my judgement

wk9874 committed:
Update test image setup script

wk9874 committed:
Fix possible index error with default image finding

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+    !!v &&
+    typeof v === "object" &&
+    "id" in v &&
+    typeof (v as { id: unknown }).id === "string"
```
The line above type-narrows in a way that we don't need to cast to `{ id: unknown }` - `v.id` is guarenteed

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/b952446209ef657500bb82764ffc38515a45e191

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+  off?: (event: string, cb: (...args: unknown[]) => void) => void;
+};
+
+function asAnnotatorApi(a: unknown): AnnotatorApi | null {
```
This slightly worries me as the logic in this function does not guarentee it is of type `AnnotatorApi` for an unknown type so this does nothing to improve the type safety. It seems like you are only using this to cast something of type `Annotator` which already satisfies the `AnnotatorApi` so could be a direct cast - I fear having this generic function would encourage people to use the it on a different type and assume it would work when it is not guarenteed.

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/254c87c4d40c34587db3fd2e462e293e76788a6c

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+  return a as AnnotatorApi;
+}
+
+function isImageAnnotation(v: unknown): v is ImageAnnotation {
```
As before, this type guard implies you can pass anything in to it and it will confirm if it is an `ImageAnnotation` but that is not the case. For example `ImageAnnotation` requires a `target` type which is not confirmed in this guard. I think we should be more accurate if we are including these sorts of guards as it makes it look like we are type safe when we are not

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/b952446209ef657500bb82764ffc38515a45e191

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+  );
+}
+
+function toAnnoList(got: unknown): ImageAnnotation[] {
```
This is a good example of why these type guards are dangerous. In this function if I pass an array that is filled with `{id: string}` objects it will output a `ImageAnnotation` array - but this list does not satisfy the type assigned to it which can lead to UI crashes if we try to access the `target` attribute

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/b952446209ef657500bb82764ffc38515a45e191

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+
+// --- Rectangle-only guard (filters out polygons or unknown shapes) ---
+function isRectangleAnno(a: unknown): boolean {
+  const sel = (a as { target?: { selector?: unknown } } | null)?.target
```
You only ever pass in `ImageAnnotation` types into this function. Why not change the input type so that you can make use of the typing e.g.:
```
function isRectangleAnno(a: ImageAnnotation): boolean {
  const sel = a.target.selector
  return sel.type === ShapeType.RECTANGLE
}
```
This seems a lot cleaner. The only risk I see is if we are being type unsafe in other places - but that should be resolved there rather than here

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/2ae3f624b9c921e163429803362044d2049c894c

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+}
+
+export const AnnoBridge = Object.assign(
+  forwardRef<
```
`forwardRef` is a deprecated approach: [react docs](https://react.dev/reference/react/forwardRef)

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/5243604ecf3e017a172c01b3c57cb7c04da03ab1 and https://github.com/ukaea/toktagger/pull/137/commits/2b7e3743c05b3f39eadf2212d9f3500100a7ae91

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+
+    const sig = (anns: ImageAnnotation[]) =>
+      JSON.stringify(
+        anns.map((a) => ({ id: a.id, t: a.type, tSrc: a.target?.source })),
```
The typing is off here. I think it stems from the unsafe type assertions leading to it looking like you are passing in an `ImageAnnotation` array when you actually are not. The solution I think is to add additional typing

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/4a2748f0b31473609683e987df1b365e18883f9f

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+    // Double-RAF to ensure Annotorious has settled before re-enabling event handling
+    const doubleRAF = useCallback(async () => {
+      await new Promise<void>((r) =>
+        requestAnimationFrame(() => requestAnimationFrame(() => r())),
```
I'm not a huge fan of this, it seems like a bit of a code smell however I understand sometimes things like this are necessary

jblake42 replied:
Will see if it is needed after refactor

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+      if (batch && anno) {
+        const s = sig(batch);
+        if (s !== lastAppliedRef.current) {
+          const api = asAnnotatorApi(anno);
```
What is the reason for this cast? anno is of type `Annotator` which has the funciton you want to call? e.g.:
```
if (batch && anno) {
        const s = sig(batch);
        if (s !== lastAppliedRef.current) {
          await runSilently(() => anno.setAnnotations(batch, true))
          lastAppliedRef.current = s;
        }
      }
```
I can't see where a typing or linting error would come from? If this is possible I would prefer it as it avoids the need for the unsafe type guard

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/2e4200314f12ff8f3863c98b8c586ba59483d7b4

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+    const normalizeWrite = useCallback(
+      (raw: ImageAnnotation[]) => {
+        // Rectangle-only: drop anything not a rectangle
+        const rects = (raw || []).filter(isRectangleAnno);
```
If `raw` is of type `ImageAnnotation[]` why is the `|| []` needed?

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/b72ca0357cb7449d0f1411cf025a4e85d806b054

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+              [keyLower]: {
+                id: String(fixedId),
+                name: selectedClass,
+                profileId,
```
The typing of this is wrong

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/964a9739e1097520a5593a94050fa305f0c3c3f8

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+        isAnnotatorReady: () => !!anno,
+
+        persistWorkingNow: async (currentKey: string) => {
+          const api = asAnnotatorApi(anno);
```
What is the reason for this cast when all you do is call a funciton that is already contained within the type?

abdullah-ukaea replied:
Thanks for spotting this. Since  useAnnotator() already returns an Annotator | null,  the isAnnotatorApi check here isn’t doing anything. I’ll simplify this to a plain null-check and call getAnnotations directly.

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/402aa726de9b32c0db5f5eb0d1ccdd533fbdbab1

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+          if (!api) return;
+
+          await runSilently(() => {
+            api.setAnnotations?.([], true);
```
Same question, surely this method call is contained within `anno` without the type cast?

abdullah-ukaea replied:
Yep I agreed. useAnnotator() already gives us an Annotator, so the extra isAnnotatorApi check here is redundant; we can just null-check anno and call setAnnotations directly.

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/ae3eef1d069e564924744fe80149b1c0dae9c8b8

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+
+          currentKeyRef.current = currentKey;
+
+          const src = Array.isArray(list) ? list.filter(isRectangleAnno) : [];
```
`list` is of type `ImageAnnotation[]` is it not guarenteed to be an array?

abdullah-ukaea replied:
yep, list is already typed as ImageAnnotation[], so the Array.isArray guard is redundant. I’ll simplify it to list.filter() to match the type

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/07031bdccea16dd6ba454ec830099ac86522bb23

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+          const src = Array.isArray(list) ? list.filter(isRectangleAnno) : [];
+          const stamped: ImageAnnotation[] = src.map((a) => ({
+            ...deepClone(a),
+            target: { ...(a.target ?? {}), source: currentKey },
```
`a.target` should always be defined - if we need this check it is because we have been type unsafe

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/6cca80340bd70c3e40fc6ab50f515c76fbf56d6b

jblake42 reviewed (commented):
Sorry, still working my way through

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+      [getSelectedClassName, getSelectedProfile],
+    );
+
+    useImperativeHandle(
```
The use of this hook is interesting. There are a few places where non-imperitive behaviour is being carried out - specifically the getters towards the end. It is preferred where possible to pass declaritive information like this through props. This may not be necessary but it is something to consider

abdullah-ukaea replied:
Will try to address in the refactor PR

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+        dirtyRef.current = true;
+
+        const api = asAnnotatorApi(anno);
+        const rawFull = toAnnoList(api?.getAnnotations?.());
```
From inspection I don't see why this wouldn't work:
```
const rawFull = toAnnoList(anno.getAnnotations());
```
Removing the need for the guard

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+          const firstByInstance = new Map<string, string>();
+          let duplicateAnno: ImageAnnotation | null = null;
+
+          for (const a of normalized) {
```
There is a lot of use of single letter variable names in fair complex situations, I think using more explanatory names would be beneficial throughout this PR. Things like `a` and `aid` are quite opaque from inspection

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/aa3efc120a01410d3f3bcfeb0773d1bdde7e1676

jblake42 commented on toktagger/ui/src/app/frames/components/bridge.tsx:
```diff
@@ -0,0 +1,454 @@
+      };
+
+      const api = asAnnotatorApi(anno);
+      api?.on?.("createAnnotation", onCreate);
```
This is not type safe which is why TypeScript complains about this line and the destructor below. Based on the event definition any number of unknown inputs can be provided to the event, however the callback only accepts a specific type of parameter. Whilst this may be the only thing it can receive, the typing does not show this.

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+/** ------------------------------------------------------------------
+ *  Toolbar — simple frame navigation controls (Prev/Next/Jump)
+ *  ------------------------------------------------------------------ */
+export function Toolbar({
```
This doesn't appear to be used?

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/3233ea60c3dd3d5aac37a39cffa37e11576221f8

jblake42 reviewed (commented):
Another day, another small number of reviews

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+        onSelectionChange={(key) =>
+          setSelectedClassName((key as string) || null)
+        }
+        placeholder="— Select class —"
```
I am in two minds between just defaulting to the top class as I worry it will confuse users when they cannot draw onto the frame. It is either that or we should have a toast that warns them that no class is selected

abdullah-ukaea replied:
I agree about defaulting to the top class to make the user experience better, should I add this now?

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+        {LABEL_MAP.categories.map((c) => (
+          <Item key={c.name}>{c.name}</Item>
+        ))}
+      </Picker>
```
Spectrum docs seem to prefer this approach for clarity:
```
<Picker
     aria-label="Class"
     items={LABEL_MAP.categories}
     selectedKey={selectedClassName}
     onSelectionChange={(key) =>
          setSelectedClassName((key as string) || null)
     }
     placeholder="Select class"
     width="100%"
>
     {item => <Item key={item.name}>{item.name}</Item>}
</Picker>
```
I'm happy if you think the other approach is cleaner

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/45dd6b5c4ef8fb7b5826e4df040b7e430b5ca4f5

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+        placeholder="— Select class —"
+        width="100%"
+      >
+        {LABEL_MAP.categories.map((c) => (
```
I worry a little about the disconnect between this component and its parent due to you directly referencing the label map here. If for example it is changed so that the parent pulls data from an API to fetch the mapping this logic would break. Would it make more sense to pass the items as a prop into this component?

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/a6d5cb3d51189dbdaae2ed33bef4b00f79004a02

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+/** Profile type used by InstancePanel.
+ *  NOTE: This is a UI-level type describing per-instance tracking profiles.
+ */
+export type Profile = {
```
Does this need to be exported? I can't see it being used anywhere else? Minimising exports helps to avoid conflicts

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/67c87f8dace1b603888d79ac52ab327fec49d4d1

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+  );
+  const [trackId, setTrackId] = useState<string>("");
+
+  useEffect(() => {
```
`useEffect` for pure react state-based logic can be a code smell. I think this could easily be handled in the event handler for the `Add Profile` button.

abdullah-ukaea replied:
I agree, addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/96dd0fe9f3f9855e696214ef431d96ab645d4025

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+  const [className, setClassName] = useState<string>(
+    LABEL_MAP.categories[0].name,
+  );
+  const [trackId, setTrackId] = useState<string>("");
```
Is it safer to initialise this to a valid track id rather than just blank - it is likely a non-issue but it seems like a more defensive approach

abdullah-ukaea replied:
Addressed in commit  https://github.com/ukaea/toktagger/pull/137/commits/aaff5089098b22e74d6845e4698394c6b7158bac

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+/** ------------------------------------------------------------------
+ *  ConfirmModal — generic confirm dialog used by destructive flows
+ *  ------------------------------------------------------------------ */
+export function ConfirmModal({
```
Nice abstracted component

jblake42 commented on toktagger/ui/src/app/frames/components/ui.tsx:
```diff
@@ -0,0 +1,432 @@
+  const handleDelete = async () => {
+    await anno?.removeAnnotation?.(annotation.id);
+    // Two RAFs to ensure Annotorious state has settled before we notify
+    await new Promise<void>((r) =>
```
I have tested without this delay - what behaviour requires these because I couldn't see any issues with it? Is it a rare race condition?

abdullah-ukaea replied:
Thanks for spotting this. I played around with this and couldn’t reproduce any state issues so I removed the double RAF.

Addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/e993d2fb95c37f1dab6e285965e0c83fb2ecc896

jblake42 commented on toktagger/ui/src/app/projects/project_id/samples/sample_id/page.tsx:
I think a lot of the need for these guards and casts to unknown types stems from the fact we are incorrectly casting to other types elsewhere. I don't think this probably stems from just this PR but this is the reason I try and push back against casting unless it is absolutely necessary as it causes issues later down the line and removes the inherent safety that comes with typescript. I think we need to look to refactor when possible to move back to type safety and remove the needs for these guards.

abdullah-ukaea replied:
Good points in commit https://github.com/ukaea/toktagger/commit/a7bbeb57faa7903b45e6315aa890ca34ad2f7b77 I tried to reduce the need for casts/guards by aligning the frontend with the actual backend contract. For UFO I removed the frame-bounds typing/parsing because the backend doesn’t return min/max today, and kept frame: null so the backend remains the source of truth for initial frame selection. Happy to reintroduce strict bounds/types once the API exposes them so we can remove the remaining defensive glue.

jblake42 commented on toktagger/ui/src/app/projects/project_id/samples/sample_id/page.tsx:
```diff
@@ -174,76 +228,255 @@ export default function SamplePage() {
+  });
+
+  // ------------------------------
+  // UFO frame navigation driver
```
It seems strange to me that this specific frame navigation logic is in this generic section. This sort of navigation logic is surely similar to the navigation logic of other sample types which are able to move the logic within the sample specific code. Do you think it would be possible to do a similar thing for the frame code or is there a specific requirement for it to be here?

abdullah-ukaea replied:
Addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/192f236dd25c9930997f609d88935a770c820ad7

You are right, SamplePage should stay task-agnostic so I moved the UFO-only navigation into the UFO module ufoview in frames.tsx. Kept backend-driven initial frame selection in SamplePage to preserve existing fetch semantics.

jblake42 commented on toktagger/ui/src/app/frames/components/frames.tsx:
```diff
@@ -0,0 +1,1436 @@
+    ufoHasUnsavedChanges?: () => boolean;
+    ufoMarkSaved?: () => void;
+
+    ufoCollectForSave?: () => Promise<ImageAnnotation[]>;
```
Another good example of why global variables can be dangerous. This appears to be defined in two places with different typing

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/cbe807fb99c33af4cb20634534474834c73c83a7

jblake42 commented on toktagger/ui/src/app/frames/components/frames.tsx:
```diff
@@ -0,0 +1,1436 @@
+      return;
+    }
+
+    if (/^[0-9]*$/.test(newValue)) {
```
There appears to be some redundancy here. The regex and `isFinite` check seem to do quite similar things. I personally would get rid of the regex and use `isInteger` to drive the `onJump` call or generate an error message

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/65c1363d11e3c93ae438ba53e3d70c049e6a9bd8

jblake42 commented on toktagger/ui/src/app/frames/components/frames.tsx:
```diff
@@ -0,0 +1,1436 @@
+    setToastMessage(msg);
+    setToastOpen(true);
+
+    if (typeof window !== "undefined") {
```
Do you need the window, can you not just use?:
```
setTimeout(() => setToastOpen(false), 2000)
```

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/899dfb78ca57567e1d4068b785adf730bd210cf3

jblake42 commented on toktagger/ui/src/app/frames/components/frames.tsx:
```diff
@@ -0,0 +1,1436 @@
+  const drawingEnabled = !!getSelectedProfile() || !!getSelectedClassName();
+
+  const frameNumber: number = typeof data.frame === "number" ? data.frame : 0;
+  const frameLabel = Number.isFinite(frameNumber) ? frameNumber : "?";
```
This dual level type checking to confirm something that is typed as `number` is actually a number is either not necessary or a result of dodgy type assertion elsewhere

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/4eb4c7b37e5a8d7cd572081ec134fdd92c720f48

jblake42 commented on toktagger/ui/src/app/frames/components/lib.ts:
```diff
@@ -0,0 +1,1214 @@
+
+  const allKeys = Object.keys(storage);
+  const keys = options.keyPrefix
+    ? allKeys.filter((k) => k.startsWith(options.keyPrefix))
```
Typescript complains about this due to type narrowing across closures. To get rid of this a better approach is:
```
const prefix = options.keyPrefix
  const keys = prefix
    ? allKeys.filter((k) => k.startsWith(prefix))
    : allKeys;
```

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/13a5f28844fc5bee438821de0528ba1bf5f188eb

jblake42 commented on toktagger/ui/src/app/frames/components/lib.ts:
```diff
@@ -0,0 +1,1214 @@
+    bodyOut.push(cb);
+  }
+
+  return {
```
This is the sort of type casting which I believe is causing the need to do type guarding in other places. It would be nice to try and refactor this

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/673d920f7ba5ac4f99fd299a052f278d1476e96d

jblake42 commented on toktagger/ui/src/app/frames/components/lib.ts:
```diff
@@ -0,0 +1,1214 @@
+    bodyOut.push(cb);
+  }
+
+  return {
```
...and here

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/fa18f13c01bbf0f06b22598795082500b0877725

jblake42 commented on toktagger/ui/src/app/frames/components/lib.ts:
```diff
@@ -0,0 +1,1214 @@
+  const byFrame: Record<number, CocoFrame> = {};
+
+  for (const a of Array.isArray(list) ? list : []) {
+    const src = a?.target?.source;
```
The typing is weird here again, `source` does not exist in `target`

abdullah-ukaea replied:
addressed in commit https://github.com/ukaea/toktagger/pull/137/commits/61f34ce29bc2e273df56be2fe5b7ead5247cec54

jblake42 reviewed (changes requested):
This is really good work to get this tool working to this level. From testing it performs really well. It was a big PR so there are a reasonable number of comments which is to be expected. I do have a couple of fears around the approach which might be unjustified. It is very complicated and makes use of approaches which the rest of the codebase does not use such as browser local storage, custom events and global variables. This makes things quite complex and I worry what it will do for maintainability in the future. From my initial interpretation of this (which is limited as I have only really looked at it for a couple of days) I believe that use of a similar approach to other tooling using contexts would work and would bring it into alignment with the rest of the code and perhaps simplify things. I think it would be worth a discussion about whether we could refactor at some point to try and simplify.
This doesn't take away from the fact this is very impressive and will help us start to prove we can annotate frames effectively. Well done!

abdullah-ukaea committed:
Fix TS keyPrefix narrowing warning by capturing options.keyPrefix in a local prefix const before filtering localStorage keys. No functional change—still scans the same keys and produces the same counts.

abdullah-ukaea committed:
Refactor writeClassAndTrack to return { ...a, body, bodies } instead of casting an unknown object to ImageAnnotation. This removes the unsafe type assertions that caused TS warnings and reduces downstream type-guarding, while keeping the runtime behavior the same.

abdullah-ukaea committed:
Update writeClassAndTrack to consistently write class + track_id into W3C bodies (and mirror into legacy body) by upgrading plain-string labels to a structured, JSON-encoded payload. Refactor extractClassLabel and extractClassLabelFromAnnotation to reliably read class data from both the new encoded format and older back-compat formats. Introduce ClassValuePayload plus helpers makeBodyId, makeClassBody, encodeClassValue, and decodeClassValue, and add the AnnotationBody import to align with Annotorious typing and reduce parsing/format drift across the codebase.

abdullah-ukaea committed:
Refactor writeClassOnly to match the type-safe class payload flow used in writeClassAndTrack. We now encode the class payload into a JSON string for AnnotationBody.value and update existing bodies by decoding/re-encoding instead of spreading unknown values. This removes the TS warnings while keeping the same behavior and preserving the legacy body field for back-compat.

abdullah-ukaea committed:
Refactor w3cToCocoFrames to stop reading a.target.source directly, since source isn’t part of the typed ImageAnnotation.target shape in @annotorious/react. Added a small getTargetSource helper that safely extracts the W3C target.source string via runtime guards. This keeps the runtime behaviour identical while removing the typing mismatch and clarifying intent.

abdullah-ukaea committed:
linting prettier changes

abdullah-ukaea committed:
Refactor bridge.tsx to remove the unsafe asAnnotatorApi cast, which only asserted types without guaranteeing the runtime shape.
Add isFunction and a real isAnnotatorApi type guard that validates the optional Annotorious API methods are functions when present.
Update all call sites to use the guard and then treat anno as the typed API, improving type safety without changing behaviour.

abdullah-ukaea committed:
Tighten isImageAnnotation to be an accurate type guard by validating the minimum ImageAnnotation shape we actually depend on (string id plus target.selector). Removed the unnecessary id cast after the id in v check and switched to safe Record<string, unknown> property access. This makes toAnnoList filtering genuinely type-safe without changing runtime behaviour.

abdullah-ukaea committed:
Refined isRectangleAnno to accept ImageAnnotation instead of unknown, matching how it’s actually used throughout the bridge. This removes unnecessary casts and allows cleaner selector access while preserving the same rectangle-detection logic (RECTANGLE selector type or xywh= selector value).

abdullah-ukaea committed:
Replaced forwardRef usage in AnnoBridge with the React 19 ref-as-prop pattern to match the updated React docs. AnnoBridge now accepts ref as a normal prop and wires it into useImperativeHandle, keeping the exact same imperative API/behaviour.

abdullah-ukaea committed:
Updated sig to use a dedicated lightweight AnnoSigView type instead of ImageAnnotation. This matches what the function actually reads (id, optional type, and optional target.source) and removes unsafe assumptions/IDE warnings without changing the signature behaviour.

abdullah-ukaea committed:
Removed the isAnnotatorApi check from flushOverlay and call anno.setAnnotations directly, since useAnnotator() already returns an annotator with that API. Updated the overlay signature helper to accept ImageAnnotation[] and only hash typed fields (id + selector) to avoid relying on non-existent type/target.source properties.

abdullah-ukaea committed:
Removed the redundant raw || [] fallback in normalizeWrite since raw is already typed as ImageAnnotation[] and should never be null/undefined

abdullah-ukaea committed:
Fix class registry typing in handleCreate by storing a string profileId instead of the full selected profile object. We now derive profileId from the selected profile (prefer track_id, fallback to class_id) to match ClassDef.profileId’s string | null type.

abdullah-ukaea committed:
Replace the isAnnotatorApi guard in persistWorkingNow with a simple anno null-check since useAnnotator() already returns the correct Annotator type.

abdullah-ukaea committed:
Remove redundant isAnnotatorApi guard in clearOverlaySilently and call anno.setAnnotations directly after a null-check.

abdullah-ukaea committed:
Remove the redundant Array.isArray check in hydrateOverlay and filter rectangles directly from the typed ImageAnnotation[] list.

abdullah-ukaea committed:
updated the hydrateOverlay mapping to drop the fallback and spread a.target directly

abdullah-ukaea committed:
deleted unused type

abdullah-ukaea committed:
forgot to use bridgeprops type before in annobridge, so now delete the inline type and used this defined type instead

abdullah-ukaea committed:
return false instead of undefined in guard

abdullah-ukaea committed:
Normalize Annotorious createAnnotation handler to accept unknown and narrow via isImageAnnotation before running async logic, fixing on/off overload type errors

abdullah-ukaea committed:
prettier linting changes

abdullah-ukaea committed:
renamed some opaque variables e.g. a, a.id in loops to more descriptive ones. and did some prettier formatting after

abdullah-ukaea committed:
Align window.ufoCollectForSave global typing in frames.tsx to Promise<unknown> to match toolbar.tsx

abdullah-ukaea committed:
Simplify FrameSearch validation by removing the digit regex and using Number.isInteger for jump-to-frame input handling.

abdullah-ukaea committed:
removed the unnecessary window guard by using the global setTimeout

abdullah-ukaea committed:
Simplify frame number handling by using data.frame directly while keeping the finite-value guard for the displayed label.

abdullah-ukaea committed:
add vendor-css.d.ts for third-party CSS imports. Removes TS(2882) warnings for Annotorious and react-contexify styles used in frames.tsx.

abdullah-ukaea committed:
removed comments that were left from debugging

abdullah-ukaea committed:
Removed the redundant inner UfoStateDetail type assertion in the ufo:state event handler.

abdullah-ukaea committed:
remove guard with selectedclassname in d

abdullah-ukaea committed:
added comments to help readability and changed 1 line as there was a Unnecessary wrapper in onUpdate

abdullah-ukaea committed:
remoeved double cast type

abdullah-ukaea committed:
Guard UFO save against missing project/sample IDs. Show a toast and bail early instead of calling saveAnnotations with undefined params.

abdullah-ukaea committed:
Add early guards in handleUfoNextSample and UfoShotSearch to prevent calling save/navigation APIs when project_id or sample_id is missing. This shows a toast instead of passing undefined IDs into backend requests.

abdullah-ukaea committed:
Guard non-UFO controls behind project_id/sample_id checks so we don’t pass undefined IDs into SaveButton, NextButton, or ShotSearch. Keep Clear always available since it doesn’t require backend IDs.

abdullah-ukaea committed:
Hoist UfoShotSearch out of ToolBar and pass the required deps as props so the component identity is stable across parent re-renders. This prevents React from unmounting/remounting the search field (triggered by the 1s instanceCounts updates), which was resetting local state and clearing the input while typing.

abdullah-ukaea committed:
Fix Spectrum button props for UFO toolbar controls
Add explicit variant + outline styling to the Rectangle placeholder button and align Clear Current Frame styling to use outlined Spectrum buttons.

abdullah-ukaea committed:
Remove the success toast for Clear Current Frame since the UI already shows immediate feedback, while keeping the error toast for failures.

abdullah-ukaea committed:
Simplify UFO frame navigation in SamplePage by removing the unused frame bounds state/effect and the associated clamp/next/prev available frames logic that relied on optional frame_bounds metadata. Frame navigation now derives currentFrame primarily from the backend-returned ImageData.frame, with a small best-effort fallback to dataParams.frame. We also keep the existing backend-driven initialization behavior ({ name: image, frame: null }) so the server continues to choose the first available frame on initial load.

abdullah-ukaea committed:
Move UFO frame navigation out of the generic SamplePage and into UFOView in frames.tsx. SamplePage now stays task-agnostic except for the existing UFO initialization (frame: null) needed for backend-selected first frame. Prev/Next/Jump still call setDataParams({ name: image, frame: ... }) exactly as before, so UI behavior remains unchanged while keeping UFO-specific logic contained in the UFO view.

abdullah-ukaea committed:
Delete the unused Toolbar export + clean up now-unused imports

abdullah-ukaea committed:
Refactor UFO ClassPanel Picker to use Spectrum’s items render-prop pattern for clearer, more declarative options rendering while preserving identical selection behavior.

abdullah-ukaea committed:
Decoupled ClassPanel from LABEL_MAP by passing class options in as an items prop instead of referencing the label map inside the component. Updated the toolbar call site accordingly so the dropdown stays driven by the parent

abdullah-ukaea committed:
Make Profile type in ui.tsx non-exported to avoid unnecessary export conflicts.

abdullah-ukaea committed:
Move auto track ID generation out of useEffect and into the “Add Profile” button handler. This keeps the same behavior while avoiding effect-driven state updates

abdullah-ukaea committed:
Initialize trackId with a valid auto-generated value (lazy initializer) instead of an empty string. This is a defensive tweak to avoid transient invalid state while keeping the UI behavior unchanged.

abdullah-ukaea committed:
Remove the extra requestAnimationFrame delay after removeAnnotation in handleDelete. Simplifies the delete flow since no race condition or UI issues were observed without it.

abdullah-ukaea committed:
Prettier linting

abdullah-ukaea committed:
fixed merge conflict related to toktagger/api/static/index.html script, using whats on main branch as source of truth

abdullah-ukaea committed:
Resolve toolbar merge: keep main tools for ELM/MHD and preserve UFO frame annotation toolbar

This resolves the merge conflict in the UI toolbar by separating non-UFO tooling from the new UFO frame annotation tooling.

Non-UFO (main behavior):
- Keep the original annotation save flow that marks annotations as validated=true.
- Keep the original save endpoint that uses NEXT_PUBLIC_API_URL + /backend-api/...
- Keep existing controls and toolbox tools for ELM/MHD (model prediction, peak/outlier/changepoint/jump, spectrogram tools).
- Keep the existing “Jump to Shot” logic that saves validated annotations before navigating.

UFO (frame annotation behavior):
- Add a dedicated UFO toolbar layout with Save/Next and shot jump that saves the current frame annotations first.
- Collect W3C frame annotations from the frame view, convert them to COCO frame format, then convert to backend video bounding boxes.
- Save the UFO payload to the UFO backend annotations endpoint, and clear the UFO working-dirty flag after successful save/navigation.
- Keep toolbar <-> FrameView synchronisation via ufo:state events (profiles, selection, registry, per-instance counts).
- Keep per-instance usage scanning to update instance count badges while editing.

Result: non-UFO functionality matches main, while UFO frame annotation tooling remains intact and unchanged in behaviour.

abdullah-ukaea committed:
Resolve merge conflict in sample page: keep main non-UFO UI, keep UFO frame annotator flow

This commit resolves the merge conflict in toktagger/ui/src/app/projects/project_id/samples/sample_id/page.tsx

Non-UFO behaviour (kept as main branch source of truth):
- Restored and kept the ModelTrainModal and ModelPredictModal in the page layout.
- Preserved the existing SampleView routing for disruption / ELM / MHD views.
- Kept the overall layout structure: ToolBar on the left, view content on the right.
- MHD still forces spectrogram view params (name=spectrogram, nperseg=256) before requesting data.

UFO behaviour (kept as abdullah/frame_annotation_v2 source of truth):
- Switched UFO rendering to use the new frame annotation UI:
  import UFOView from @/app/frames/components/frames.
- Passed through the required props for frame annotation:
  annotations, setAnnotations, dataParams/setDataParams, projectId, sampleId.
- Preserved backend-driven initial frame selection by setting:
  dataParams = { name: image, frame: null } on first load of a UFO project.
  This keeps the old “backend picks the initial frame” behaviour while enabling tooling.

abdullah-ukaea committed:
Prettier linting

abdullah-ukaea committed:
Refactor ToolBar to split UFO and non-UFO logic into separate components so React hooks are never called conditionally. Also replace any window casts with proper Window typings while keeping behaviour the same.

abdullah-ukaea committed:
Merge branch 'main' into abdullah/frame_annotation_v2

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
added 4th positional arg query_strategy to line project_id = create_project()

abdullah-ukaea committed:
Rename UFO-prefixed functions/components to video/frame terminology across the UI. Updated toolbar.tsx, frames.tsx, and page.tsx to use the new names, and renamed working-dirty helpers in adapters.ts. No functional changes—this is a semantics/naming cleanup.

abdullah-ukaea committed:
Improved documentation / comments in every video annotation tool related file

abdullah-ukaea committed:
Prettier Linting

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Refactor AnnoBridge to stop duck-typing the Annotorious annotator instance. We removed the custom AnnotatorApi shape + isAnnotatorApi guard and instead type the annotator directly using the core Annotator<Annotation, ImageAnnotation> interface returned by useAnnotator(). This lets us call getAnnotations/setAnnotations/on/off without optional chaining or misleading runtime checks, while keeping runtime validation only for truly unknown data (e.g. objects read from storage).

abdullah-ukaea committed:
Import ShapeType from Annotorious and simplify isRectangleAnno to a single ShapeType.RECTANGLE enum check. Removes the old selector value/regex probing in favour of the library’s typed Shape API.

abdullah-ukaea committed:
Tighten Annotorious annotation typing by casting the hook instance to a typed Annotator and making isRectangleAnno a proper type guard. This removes the unknown inference from getAnnotations() so .filter/.map narrow correctly,

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
merged main

jblake42 reviewed (approved):
I am happy for this to be merged based on the functionality side of it and with the caveat that we should look to refactor the implementation in order to simplify

samueljackson92 committed:
Merge branch 'main' into abdullah/frame_annotation_v2

Commit:
chore: update build output [skip ci]

samueljackson92 reviewed (changes requested):
@abdullah-ukaea I am not going to test as deeply as @jblake42 as I trust he has reviewed the details (probably better than I would). However, one quick thing I've noticed is that the images (on my screen) are very small and there's no way to zoom. 

Zooming might be beyond this PR, but perhaps we can make the size of the images reactive to the users window size? I cannot really see the UFOs to be able to tag them with the current implementation as the image is so tiny.

<img width="1915" height="886" alt="image" src="https://github.com/user-attachments/assets/628f6d98-a5cf-4e2d-90b7-5283652ce8d6" />

abdullah-ukaea committed:
Merge branch 'main' into abdullah/frame_annotation_v2

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
updated setup.py to use video task instead of ufo task

abdullah-ukaea committed:
Update the import list (add ImageData, ImageDataSchema), Update parseData(...) to handle TaskType.Video and Ensure video projects POST the right dataParams (name: image, frame)

abdullah-ukaea committed:
Added support for video samples in the sample page by rendering the existing VideoView from the frames tooling when project.task === TaskType.Video. Introduced a small VideoViewWrapperFromContext that pulls data/annotations/dataParams from useSample()

abdullah-ukaea committed:
Updated the main ToolBar to detect TaskType.Video and render VideoToolbar directly

abdullah-ukaea committed:
Keep the video frame annotator mounted while loading the next frame to prevent toolbar/session state from resetting. Move video handling ahead of the isLoading gate and only show LoadingView for video when data is missing on initial load.

abdullah-ukaea committed:
updated comments

abdullah-ukaea committed:
fixed comments, made eslint changes

abdullah-ukaea committed:
Prettier linter changes

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Measure the full available viewport width (avoid shrink-to-fit inline-flex measurements) and compute a display width that only upscales small frames toward a comfortable tagging size while keeping large frames behaving as before. Apply the computed width to the <img> so MAST frames render larger without unintentionally downscaling JET frames.

abdullah-ukaea committed:
Prettier lintign

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Merge branch 'main' into abdullah/frame_annotation_v2

Commit:
chore: update build output [skip ci]

abdullah-ukaea committed:
Video: prevent missing-frame fetch errors from kicking the user out of the annotator by treating them as navigation boundaries. We now toast + roll back dataParams.frame to the last successfully loaded frame instead of setting error/clearing data.

abdullah-ukaea committed:
Video: prevent missing-frame fetch errors from kicking the user out of the annotator by treating them as navigation boundaries. We now toast + roll back dataParams.frame to the last successfully loaded frame instead of setting error/clearing data.

abdullah-ukaea committed:
prettier linting

Commit:
chore: update build output [skip ci]
